### PR TITLE
MM-31126 Collapsed threads support

### PR DESCRIPTION
**Pull request by reflog**

### CODE REVIEW

File: src/types/users.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export type UserNotifyProps = {
 };
 
 export type UserProfile = {
-    id: string;
+    id: 'me' | string;
```
Comment (jespino) [inline src/types/users.ts:26]: Is there any reason to have this distinction here?

Comment (reflog) [reply src/types/users.ts:26]: nope. will remove

File: src/actions/threads.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                threads: userThreadList.threads.map((thread) => ({...thread, is_following: true})),
+                team_id: teamId,
+            },
+        });
+
+        return {data: userThreadList};
+    };
+}
+
+export function markThreadsRead(userId: string, teamId: string) {
```
Comment (jespino) [inline src/actions/threads.ts:46]: maybe something like `markAllTeamThreadsAsRead` from the parameters you can infer that, but is not obvious enough that you are marking as read all the team threads.

File: src/actions/threads.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        try {
+            await Client4.updateThreadsReadForUser(userId, teamId);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: ThreadTypes.THREADS_READ,
```
Comment (jespino) [inline src/actions/threads.ts:57]: Same with this, this action id is not obvious enough. something like `ALL_TEAM_THREADS_READ` would be more explicit.

File: src/reducers/entities/threads.test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {ThreadTypes} from 'action_types';
+import deepFreeze from 'utils/deep_freeze';
+import threadsReducer from './threads';
+
+describe('threads', () => {
+    describe('RECEIVED_THREADS', () => {
```
Comment (jespino) [inline src/reducers/entities/threads.test.js:9]: Missing a the tests here around the other events that can happen on the threads reducers, like the `FOLLOW_CHANGED_THREAD` and the `LEAVE_TEAM`.

And missing tests for the other existing reducers like `threadsReducer`,  `threadsInTeamReducer` and `countsReducer`.

File: src/reducers/entities/threads.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        const {threads} = action.data;
+        return {
+            ...state,
+            ...threads.reduce((results: IDMappedObjects<UserThread>, thread: UserThread) => {
+                results[thread.id] = thread;
+                return results;
+            }, {}),
+        };
+    }
+    case ThreadTypes.FOLLOW_CHANGED_THREAD: {
```
Comment (jespino) [inline src/reducers/entities/threads.ts:23]: What happens if the followed thread is not yet in the store?

File: src/selectors/entities/preferences.ts [hunk 6]
```diff
@@ -8,7 +8,7 @@ import {General, Preferences} from '../../constants';
 import {getConfig, getLicense} from 'selectors/entities/general';
 import {getCurrentTeamId} from 'selectors/entities/teams';
 
-import {PreferenceType} from 'types/preferences';
+import {PreferenceType, Theme} from 'types/preferences';
```
Comment (jespino) [inline src/selectors/entities/preferences.ts:11]: Why this changes appear here? Looks like a merge conflict that is resolved reverting changes.

**Review by jespino — CHANGES_REQUESTED** [review]
Some improvement suggestions and a doubt about a possible badly resolved merge conflict.

File: src/selectors/entities/preferences.ts (hunk 6 above)
Comment (calebroseland) [reply src/selectors/entities/preferences.ts:11]: @jespino I'm using the `getTheme` selector in a small part the main feature, and I wanted to cleanup the types around it as it's currently typed as returning `any`. Having clearer types during development helps me to be more sure, but by all means, it could be separated in another PR.

File: src/constants/preferences.ts [hunk 7]
```diff
@@ -2,9 +2,8 @@
 // See LICENSE.txt for license information.
 
 import {Theme} from 'types/preferences';
-import {Dictionary} from 'types/utilities';
 
-const Preferences: Dictionary<any> = {
+const Preferences = {
```
Comment (calebroseland) [inline src/constants/preferences.ts:6]: @jespino this is also related to that (`getTheme` cleanup)

File: src/types/threads.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    total_unread_mentions: number;
+    threads: UserThread[];
+}
+
+export type ThreadsState = {
+    threadsInTeam: RelationOneToMany<Team, UserThread>;
+    threads: IDMappedObjects<UserThread>;
+    counts: RelationOneToOne<Team, {
+        total: number;
+        total_unread_replies: number;
```
Comment (calebroseland) [inline src/types/threads.ts:32]: @reflog I spoke with Matt on this, and indeed this will be used for "You have N threads with unread messages", so we may want to rename this after it's changed server-side.

File: src/actions/threads.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: ThreadTypes.READ_CHANGED_THREAD,
+            data: {
+                id: threadId,
+                team_id: teamId,
+                timestamp,
```
Comment (calebroseland) [inline src/actions/threads.ts:82]: Should we update `unread_replies` and `unread_mentions` here? Or are we just waiting for the websocket event to update these?

File: src/actions/threads.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        dispatch({
+            type: ThreadTypes.FOLLOW_CHANGED_THREAD,
+            data: {
+                id: threadId,
+                team_id: teamId,
+                following: newState,
+            },
+        });
```
Comment (calebroseland) [inline src/actions/threads.ts:107]: Should we dispatch this lazily and then again if need be after success, to prevent this on slow connections?
CleanShot 2021-01-11 at 08 51 26

**Review by calebroseland — COMMENTED** [review]
Just a couple questions after taking a second look at this.

File: src/actions/threads.ts (hunk 9 above)
Comment (reflog) [reply src/actions/threads.ts:82]: Yeah, I'd leave this for websocket events to keep it consistent.

File: src/actions/threads.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        let userThreadList: undefined | UserThreadList;
+
+        try {
+            userThreadList = await Client4.getUserThreads(userId, teamId, {page, pageSize: perPage, extended: true});
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
```
Comment (calebroseland) [inline src/actions/threads.ts:27]:
```suggestion
        dispatch({
            type: PostTypes.RECEIVED_PROFILES_LIST,
            data: userThreadList.threads.map(({participants: users}) => users).flat(),
        });

```

**Review by calebroseland — CHANGES_REQUESTED** [review]
One last one to help with profile/user selectors.

**Review by calebroseland — APPROVED** [review]

**Review by jespino — APPROVED** [review]
LGTM
